switch/switch.8
.TH secchan 8 "May 2008" "OpenFlow" "OpenFlow Manual"

.SH NAME
switch \- userspace implementation of OpenFlow switch

.SH SYNOPSIS
.B switch
[\fIoptions\fR]
\fB-i\fR \fInetdev\fR[\fB,\fInetdev\fR]...
\fIcontroller\fR

.SH DESCRIPTION
The \fBswitch\fR is a userspace implementation of an OpenFlow switch.
It implements all three parts of the OpenFlow switch specification: a
``flow table'' in which each flow entry is associated with an action
telling the switch how to process the flow; a ``secure channel''
connecting the switch to a remote process (a controller), allowing
commands and packets to be sent between the controller and the switch;
and an OpenFlow protocol implementation.

\fBswitch\fR monitors one or more network device interfaces,
forwarding packets between them according to the entries in the flow
table.  It also maintains a connection to an OpenFlow controller over
a TCP or SSL connection, relaying packets that do not match a flow
table entry to the controller and executing commands sent by the
controller.

For access to network devices, the switch program must normally run as
root.

The mandatory \fIcontroller\fR argument specifies how to connect to
the OpenFlow controller.  It takes one of the following forms:

.TP
\fBtcp:\fIhost\fR[\fB:\fIport\fR]
The specified TCP \fIport\fR (default: 975) on the given remote
\fIhost\fR.

.TP
\fBssl:\fIhost\fR[\fB:\fIport\fR]
The specified SSL \fIport\fR (default: 976) on the given remote
\fIhost\fR.  The \fB--private-key\fR, \fB--certificate\fR, and
\fB--ca-cert\fR options are mandatory when this form is used.

.SH OPTIONS
.TP
\fB-i\fR, \fB--interfaces=\fR\fInetdev\fR[\fB,\fInetdev\fR]...
Specifies each \fInetdev\fR (e.g., \fBeth0\fR) as a switch port.  The
specified network devices should not have any configured IP addresses.
This option may be given any number of times to specify additional
network devices.

.TP
\fB-d\fR, \fB--datapath-id=\fIdpid\fR
Specifies the OpenFlow switch ID (a 48-bit number that uniquely
identifies a controller) as \fIdpid\fR, which consists of exactly 12
hex digits.  Without this option, \fBswitch\fR picks an ID randomly.

.TP
\fB-p\fR, \fB--private-key=\fIprivkey.pem\fR
Specifies a PEM file containing the private key used as the switch's
identity for SSL connections to the controller.

.TP
\fB-c\fR, \fB--certificate=\fIcert.pem\fR
Specifies a PEM file containing a certificate, signed by the
controller's certificate authority (CA), that certifies the switch's
private key to identify a trustworthy switch.

.TP
\fB-C\fR, \fB--ca-cert=\fIcacert.pem\fR
Specifies a PEM file containing the CA certificate used to verify that
the switch is connected to a trustworthy controller.

.TP
.BR \-h ", " \-\^\-help
Prints a brief help message to the console.

.TP
.BR \-v ", " \-\^\-verbose
Prints debug messages to the console.

.TP
.BR \-V ", " \-\^\-version
Prints version information to the console.

.SH "SEE ALSO"

.BR dpctl (8),
.BR ofp-pki (8),
.BR controller (8)
.BR vlogconf (8)


switch/switch.c
/* Copyright (c) 2008 The Board of Trustees of The Leland Stanford
 * Junior University
 * 
 * We are making the OpenFlow specification and associated documentation
 * (Software) available for public use and benefit with the expectation
 * that others will use, modify and enhance the Software and contribute
 * those enhancements back to the community. However, since we would
 * like to make the Software available for broadest use, with as few
 * restrictions as possible permission is hereby granted, free of
 * charge, to any person obtaining a copy of this Software to deal in
 * the Software under the copyrights without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 * 
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT.  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * 
 * The name and trademarks of copyright holder(s) may NOT be used in
 * advertising or publicity pertaining to the Software or any
 * derivatives without specific, written prior permission.
 */

#include <errno.h>
#include <getopt.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#include "command-line.h"
#include "daemon.h"
#include "datapath.h"
#include "fault.h"
#include "openflow.h"
#include "poll-loop.h"
#include "queue.h"
#include "util.h"
#include "rconn.h"
#include "vconn.h"
#include "vconn-ssl.h"
#include "vlog-socket.h"

#define THIS_MODULE VLM_switch
#include "vlog.h"

static void parse_options(int argc, char *argv[]);
static void usage(void) NO_RETURN;

static const char *listen_vconn_name;
static struct datapath *dp;
static uint64_t dpid = UINT64_MAX;
static char *port_list;

static void add_ports(struct datapath *dp, char *port_list);

int
main(int argc, char *argv[])
{
    int error;

    set_program_name(argv[0]);
    register_fault_handlers();
    vlog_init();
    parse_options(argc, argv);

    if (argc - optind != 1) {
        fatal(0, "missing controller argument; use --help for usage");
    }

    error = dp_new(&dp, dpid, rconn_new(argv[optind], 128));
    if (listen_vconn_name) {
        struct vconn *listen_vconn;
        int retval;
        
        retval = vconn_open(listen_vconn_name, &listen_vconn);
        if (retval && retval != EAGAIN) {
            fatal(retval, "opening %s", listen_vconn_name);
        }
        if (!vconn_is_passive(listen_vconn)) {
            fatal(0, "%s is not a passive vconn", listen_vconn_name);
        }
        dp_add_listen_vconn(dp, listen_vconn);
    }
    if (error) {
        fatal(error, "could not create datapath");
    }
    if (port_list) {
        add_ports(dp, port_list); 
    }

    error = vlog_server_listen(NULL, NULL);
    if (error) {
        fatal(error, "could not listen for vlog connections");
    }

    daemonize();

    for (;;) {
        dp_run(dp);
        dp_wait(dp);
        poll_block();
    }

    return 0;
}

static void
add_ports(struct datapath *dp, char *port_list)
{
    char *port, *save_ptr;

    /* Glibc 2.7 has a bug in strtok_r when compiling with optimization that
     * can cause segfaults here:
     * http://sources.redhat.com/bugzilla/show_bug.cgi?id=5614.
     * Using ",," instead of the obvious "," works around it. */
    for (port = strtok_r(port_list, ",,", &save_ptr); port;
         port = strtok_r(NULL, ",,", &save_ptr)) {
        int error = dp_add_port(dp, port);
        if (error) {
            fatal(error, "failed to add port %s", port);
        }
    }
}

static void
parse_options(int argc, char *argv[])
{
    static struct option long_options[] = {
        {"interfaces",  required_argument, 0, 'i'},
        {"datapath-id", required_argument, 0, 'd'},
        {"listen",      required_argument, 0, 'l'},
        {"detach",      no_argument, 0, 'D'},
        {"pidfile",     optional_argument, 0, 'P'},
        {"verbose",     optional_argument, 0, 'v'},
        {"help",        no_argument, 0, 'h'},
        {"version",     no_argument, 0, 'V'},
        VCONN_SSL_LONG_OPTIONS
        {0, 0, 0, 0},
    };
    char *short_options = long_options_to_short_options(long_options);

    for (;;) {
        int indexptr;
        int c;

        c = getopt_long(argc, argv, short_options, long_options, &indexptr);
        if (c == -1) {
            break;
        }

        switch (c) {
        case 'd':
            if (strlen(optarg) != 12
                || strspn(optarg, "0123456789abcdefABCDEF") != 12) {
                fatal(0, "argument to -d or --datapath-id must be "
                      "exactly 12 hex digits");
            }
            dpid = strtoll(optarg, NULL, 16);
            if (!dpid) {
                fatal(0, "argument to -d or --datapath-id must be nonzero");
            }
            break;

        case 'h':
            usage();

        case 'V':
            printf("%s "VERSION" compiled "__DATE__" "__TIME__"\n", argv[0]);
            exit(EXIT_SUCCESS);

        case 'D':
            set_detach();
            break;

        case 'P':
            set_pidfile(optarg ? optarg : "switch.pid");
            break;

        case 'v':
            vlog_set_verbosity(optarg);
            break;

        case 'i':
            if (!port_list) {
                port_list = optarg;
            } else {
                port_list = xasprintf("%s,%s", port_list, optarg);
            }
            break;

        case 'l':
            if (listen_vconn_name) {
                fatal(0, "-l or --listen may be only specified once");
            }
            listen_vconn_name = optarg;
            break;

        VCONN_SSL_OPTION_HANDLERS

        case '?':
            exit(EXIT_FAILURE);

        default:
            abort();
        }
    }
    free(short_options);
}

static void
usage(void)
{
    printf("%s: userspace OpenFlow switch\n"
           "usage: %s [OPTIONS] CONTROLLER\n"
           "where CONTROLLER is an active OpenFlow connection method.\n",
           program_name, program_name);
    vconn_usage(true, true);
    printf("\nConfiguration options:\n"
           "  -i, --interfaces=NETDEV[,NETDEV]...\n"
           "                          add specified initial switch ports\n"
           "  -d, --datapath-id=ID    Use ID as the OpenFlow switch ID\n"
           "                          (ID must consist of 12 hex digits)\n"
           "  -l, --listen=METHOD     allow management connections on METHOD\n"
           "                          (a passive OpenFlow connection method)\n"
           "\nOther options:\n"
           "  -D, --detach            run in background as daemon\n"
           "  -P, --pidfile[=FILE]    create pidfile (default: %s/switch.pid)\n"
           "  -v, --verbose           set maximum verbosity level\n"
           "  -h, --help              display this help message\n"
           "  -V, --version           display version information\n",
        RUNDIR);
    exit(EXIT_SUCCESS);
}
